packageurl_test.go

```diff
@@ -119,150 +116,198 @@ func (t TestFixture) Qualifiers() packageurl.Qualifiers {
-		//       because qualifiers may not be in any order. By reparsing back
-		//       we can ensure the data transfers between string and instance form.
-		canonical, _ := packageurl.FromString(tc.CanonicalPurl)
-		toTest, _ := packageurl.FromString(result)
-		// If the two results don't equal then the ToString failed
-		if !reflect.DeepEqual(toTest, canonical) {
-			t.Logf("%s failed: %s != %s", tc.Description, result, tc.CanonicalPurl)
+	} else {
+		// Invalid cases
+		if err == nil {
+			t.Logf("%s did not fail and returned %#v", tc.Description, p)
 			t.Fail()
 		}
 	}
+
 }
 
-// TestStringer verifies that the Stringer implementation produces results
-// equivalent with the ToString method.
-func TestStringer(t *testing.T) {
-	// Read the json file
-	data, err := os.ReadFile("testdata/test-suite-data.json")
-	if err != nil {
-		t.Fatal(err)
+func buildTest(tc TestFixture, t *testing.T) {
+	input := tc.Input.PurlComponent
+	instance := packageurl.NewPackageURL(
+		input.PackageType, input.Namespace, input.Name, input.Version,
+		// Use QualifiersFromMap so that the qualifiers have a defined order, which is needed for string comparisons
+		packageurl.QualifiersFromMap(input.Qualifiers().Map()), input.Subpath)
+	result := instance.ToString()
+	canonicalExpectedPurl := tc.ExpectedOutput.Purl
+
+	if tc.ExpectedFailure == false {
+		if result != *canonicalExpectedPurl {
+			t.Logf("%s: '%s' test failed: wanted: '%s', got '%s'", tc.Description, tc.TestType, *canonicalExpectedPurl, result)
+			t.Fail()
+		}
+	} else {
+		t.Logf("%s did not fail and returned %#v", tc.Description, instance)
+		t.Fail()
 	}
-	// Load the json file contents into a structure
-	var testData []TestFixture
-	err = json.Unmarshal(data, &testData)
+
+}
+
+func TestPurlSpecFixtures(t *testing.T) {
+	testFiles, err := readJSONFilesFromDir("testdata/purl-spec/tests/types/")
 	if err != nil {
 		t.Fatal(err)
 	}
-	// Use ToString on each item
-	for _, tc := range testData {
-		// Skip invalid items
-		if tc.IsInvalid == true {
-			continue
-		}
-		purlPtr := packageurl.NewPackageURL(
-			tc.PackageType, tc.Namespace, tc.Name,
-			tc.Version, tc.Qualifiers(), tc.Subpath)
-		purlValue := *purlPtr
-
-		// Verify that the Stringer implementation returns a result
-		// equivalent to ToString().
-		if purlPtr.ToString() != purlPtr.String() {
-			t.Logf("%s failed: Stringer implementation differs from ToString: %s != %s", tc.Description, purlPtr.String(), purlPtr.ToString())
-			t.Fail()
+
+	for _, data := range testFiles {
+		var suite TestSuite
+		err := json.Unmarshal(data, &suite)
+		if err != nil {
+			t.Fatal(err)
 		}
 
-		// Verify that the %s format modifier works for values.
-		fmtStr := purlValue.String()
-		if fmtStr != purlPtr.String() {
-			t.Logf("%s failed: %%s format modifier does not work on values: %s != %s", tc.Description, fmtStr, purlPtr.ToString())
-			t.Fail()
+		for _, tc := range suite.Tests {
+			t.Run(tc.TestType, func(t *testing.T) {
```

Comment:
Name each test case to something like
`    --- FAIL: TestPurlSpecFixtures/npm-test.json[15]build` 
instead of 
`    --- FAIL: TestPurlSpecFixtures/build#143`
```suggestion
		for idx, tc := range suite.Tests {
			testName := fmt.Sprintf("%s[%d]%s", file.name, (idx + 1), tc.TestType)
			t.Run(testName, func(t *testing.T) {
```